Automattic/woocommerce-payments
# Fix multiple conversions with subscription switching, minor refactor for manual renewals

Fixes #2125

#### Changes proposed in this Pull Request

* Move/add filters for subscriptions to `Compatibility` class.
* Include `MultiCurrency` class into `Compatibility` class.
* `get_subscription_product_price` returns proper amount for subscription price.
* `get_subscription_product_signup_fee` returns proper amount for subscription sign up fee.
* `maybe_disable_mixed_cart` disables allowing more than just a single subscription being added to the cart. Presently only when there's a switch in the cart. Prevents subscriptions with different currencies being added to the cart.
* Update `override_selected_currency` to check for subscription switches and force the currency to the sub's currency.
* Update `should_hide_widgets` to hide widgets if a switch is happening.
* `get_subscription_switch_cart_items` returns subscription cart items.
* `get_subscription_switch_id_from_superglobal` gets sub switch id from `$_GET` superglobal.
* `get_product_price` consolidated checks. Hoped to remove compatibility here, but due to how prices are filtered, we cannot do that just yet.
* Updated `WC_Subscriptions` helper class to handle `wcs_get_order_type_cart_items` calls.
* Updated/added tests to match additions and changes.

#### Testing instructions

For sub switching:

1. Go to WooCommerce > Settings > Subscriptions and enable switching, also setting to prorate recurring payment for upgrades and downgrades.
2. Use or set up a variable sub product that has at least 2 variations with different prices, such as 10/month and 20/month.
3. Sign up for the lower price.
4. Once order is complete, scroll down to use the _View_ option to view the subscription.
5. Use the _Upgrade or Downgrade_ button on the subscription to start the process.
6. Choose the higher priced variation on the product and add to cart.
7. The cart should show the sign up fee as the proper prorated amount to upgrade.

For manual renewals:

1. Go to WooCommerce > Settings > Subscriptions and enable _Accept Early Renewal Payments_.
2. Purchase a subscription, or use existing one. 
3. Go to My Account > Subscriptions, then _View_ a subscription.
4. Click _Renew now_.
5. The cost of the subscription should be the same, meaning not double converted.

-------------------
